examples/step-44/step-44.cc

```diff
@@ -2001,16 +1998,18 @@ namespace Step44
       {
         fe_values_ref.reinit(cell);
 
-        PointHistory<dim> *lqph =
-          reinterpret_cast<PointHistory<dim>*>(cell->user_pointer());
-
-        Assert(lqph >= &quadrature_point_history.front(), ExcInternalError());
-        Assert(lqph <= &quadrature_point_history.back(), ExcInternalError());
+        // In contrast to that which was previously called for,
+        // in this instance the quadrature point data is specifically
+        // non-modifiable since we will only be accessing data.
+        const std::vector<std::shared_ptr<const PointHistory<dim> > > lqph =
+            const_cast<const CellDataStorage<typename Triangulation<dim>::cell_iterator,
+                                             PointHistory<dim> >&>(quadrature_point_history).get_data(cell);
```

Comment:
With
```c++
//    const std::vector<std::shared_ptr<const PointHistory<dim> > > lqph =
//        const_cast<const CellDataStorage<typename Triangulation<dim>::cell_iterator,
//                                         PointHistory<dim> >&>(quadrature_point_history).get_data(cell);
    const std::vector<std::shared_ptr<const PointHistory<dim> > > lqph =
        quadrature_point_history.get_data(cell);
```
in `Solid<dim>::assemble_system_tangent_one_cell`, one gets
```c++
/Users/jp/Programs/GitHub/deal.II/dealii/examples/step-44/step-44.cc:2157:67: error: no viable conversion from 'vector<shared_ptr<PointHistory<[...]>>>' to 'const vector<shared_ptr<const PointHistory<[...]>>>'
    const std::vector<std::shared_ptr<const PointHistory<dim> > > lqph =
                                                                  ^
/Users/jp/Programs/GitHub/deal.II/dealii/examples/step-44/step-44.cc:2113:29: note: in instantiation of member function 'Step44::Solid<3>::assemble_system_tangent_one_cell' requested here
                    &Solid::assemble_system_tangent_one_cell,
                            ^
```
and
```c++
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/vector:537:5: note: candidate constructor not viable: no known conversion from 'std::vector<std::shared_ptr<PointHistory<3> > >' to
      'initializer_list<value_type>' (aka 'initializer_list<std::__1::shared_ptr<const Step44::PointHistory<3> > >') for 1st argument
    vector(initializer_list<value_type> __il);
    ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/vector:549:5: note: candidate constructor not viable: no known conversion from 'std::vector<std::shared_ptr<PointHistory<3> > >' to
      'const std::__1::vector<std::__1::shared_ptr<const Step44::PointHistory<3> >, std::__1::allocator<std::__1::shared_ptr<const Step44::PointHistory<3> > > > &' for 1st argument
    vector(const vector& __x);
    ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/vector:555:5: note: candidate constructor not viable: no known conversion from 'std::vector<std::shared_ptr<PointHistory<3> > >' to
      'std::__1::vector<std::__1::shared_ptr<const Step44::PointHistory<3> >, std::__1::allocator<std::__1::shared_ptr<const Step44::PointHistory<3> > > > &&' for 1st argument
    vector(vector&& __x)
    ^
```
The problem is not that `CellDataStorage` doesn't have a `const` version of `get_data`, but rather that the compiler doesn't try to use it because, for example, `assemble_system_tangent_one_cell` is non-const and `quadrature_data` are simultaneously non-const.